R/learnerHelp.R

```diff
@@ -0,0 +1,230 @@
+        break
+      } 
+      if (inherits(next.learner, "RLearner")) {
+        messagef("The learner '%s' is a wrapped learner. Showing documentation of '%s' instead.",
+          learner$id, next.learner$id)
+        return(Recall(next.learner, param))
+      }
+      current.learner = next.learner
+    }
+    stop("The learner does not provide documentation.")
+  }
+  if (is.null(param)) {
+    if (length(learner$help.list) > 0) {
+      cat("Documentation for the following parameters is present:\n")
+      for (p in names(learner$help.list)) {
+        catf("- %s (%s)", p,
+          stri_match_first_regex(learner$help.list[[p]], "(?<= )[^ ]*::[^\\n]*"))
+      }
+    }
+    missingdoc = setdiff(getParamIds(learner$par.set), names(learner$help.list))
+    if (length(missingdoc) > 0) {
+      catf("The following parameters are missing documentation:\n%s",
+        collapse(strwrap(collapse(missingdoc, sep = ", "), indent = 2), "\n"))
+    }
+    return(invisible(NULL))
+  }
+  assertCharacter(param, any.missing = FALSE)
+  all.param = getParamIds(learner$par.set)
+  not.found.param = param[!param %in% all.param]
+  if (length(not.found.param) > 0) {
+    stopf("Parameters not found: %s", collapse(not.found.param, sep = ", "))
+  }
+  # put param in the same order as par.set
+  param = all.param[all.param %in% param]
+  values = getHyperPars(learner)
+  for (p in param) {
+    catf("  *%s*:", p)
+    print(learner$par.set$pars[[p]])
+    req = learner$par.set$pars[[p]]$requires
+    if (!is.null(req)) {
+      cat("Requires: ")
+      print(req)
+    }
+    if (!is.null(values[[p]])) {
+      catf("Value: %s",
+        ifelse(testScalar(values[[p]]), values[[p]], capture.output(print(values[[p]]))))
+    }
+    if (p %in% names(learner$help.list)) {
+      output = learner$help.list[[p]]
+      cat(collapse(strwrap(output), sep="\n"), "\n\n")
+    } else {
+      catf("No documentation found. Try to consult learnerHelp(\"%s\").\n", getLearnerId(learner))
+    }
+  }
+}
+
+# remove nesting levels of XML tags
+simplifyNode = function(node) {
+  children = XML::xmlChildren(node)
+  lens = nchar(stri_trim(vcapply(children, XML::xmlValue)))
+  if (length(lens) < 1) {
+    return(NULL)
+  }
+  if (sum(lens) == max(lens)) {
+    return(simplifyNode(children[[which.max(lens)]]))
+  } else {
+    return(children[lens != 0])
+  }
+}
+
+# collect all <li><code>xxx</code>yyy</li> in the document
+# and form a data.frame with two columns corresponding to xxx and yyy.
+codeListToTable = function(html) {
+  lis = XML::getNodeSet(html, "//li")
+  lislis = lapply(lis, function(li) {
+      lichi = simplifyNode(li)
+      if (length(lichi) < 2 || names(lichi)[1] != "code") {
+        return(NULL)
+      }
+      parname = XML::xmlValue(lichi[[1]])
+      pardesc = stri_join_list(lapply(lichi[-1], XML::xmlValue), collapse = " ")
+      stri_trim(c(parname, pardesc))
+    })
+  as.data.frame(do.call(rbind, lislis), stringsAsFactors = FALSE)
```

Comment:
instead of do.call, we could use one of: rbindlist, setDF(transpose(lislis)), setDF(lislis)